api/src/main/java/org/apache/iceberg/transforms/Timestamps.java

```diff
@@ -28,57 +29,131 @@
+          case YEARS:
+            return YEAR_FROM_NANOS;
+          case MONTHS:
+            return MONTH_FROM_NANOS;
+          case DAYS:
+            return DAY_FROM_NANOS;
+          case HOURS:
+            return HOUR_FROM_NANOS;
+        }
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported timestamp unit: " + type.unit());
+    }
+    throw new IllegalArgumentException(
+        "Unsupported source/result type units: " + type + "->" + resultTypeUnit);
+  }
 
   @Immutable
   static class Apply implements SerializableFunction<Long, Integer> {
-    private final ChronoUnit granularity;
+    private final ChronoUnit sourceTypeUnit;
+    private final ChronoUnit resultTypeUnit;
 
-    Apply(ChronoUnit granularity) {
-      this.granularity = granularity;
+    Apply(ChronoUnit sourceTypeUnit, ChronoUnit resultTypeUnit) {
+      this.sourceTypeUnit = sourceTypeUnit;
+      this.resultTypeUnit = resultTypeUnit;
     }
 
     @Override
-    public Integer apply(Long timestampMicros) {
-      if (timestampMicros == null) {
+    public Integer apply(Long timestampUnits) {
+      if (timestampUnits == null) {
         return null;
       }
 
-      switch (granularity) {
-        case YEARS:
-          return DateTimeUtil.microsToYears(timestampMicros);
-        case MONTHS:
-          return DateTimeUtil.microsToMonths(timestampMicros);
-        case DAYS:
-          return DateTimeUtil.microsToDays(timestampMicros);
-        case HOURS:
-          return DateTimeUtil.microsToHours(timestampMicros);
+      switch (sourceTypeUnit) {
+        case MICROS:
+          switch (resultTypeUnit) {
+            case YEARS:
+              return DateTimeUtil.microsToYears(timestampUnits);
+            case MONTHS:
+              return DateTimeUtil.microsToMonths(timestampUnits);
+            case DAYS:
+              return DateTimeUtil.microsToDays(timestampUnits);
+            case HOURS:
+              return DateTimeUtil.microsToHours(timestampUnits);
+            default:
+              throw new UnsupportedOperationException(
+                  "Unsupported result type unit: " + resultTypeUnit);
+          }
+        case NANOS:
+          switch (resultTypeUnit) {
+            case YEARS:
+              return DateTimeUtil.nanosToYears(timestampUnits);
+            case MONTHS:
+              return DateTimeUtil.nanosToMonths(timestampUnits);
+            case DAYS:
+              return DateTimeUtil.nanosToDays(timestampUnits);
+            case HOURS:
+              return DateTimeUtil.nanosToHours(timestampUnits);
+            default:
+              throw new UnsupportedOperationException(
+                  "Unsupported result type unit: " + resultTypeUnit);
+          }
         default:
-          throw new UnsupportedOperationException("Unsupported time unit: " + granularity);
+          throw new UnsupportedOperationException(
+              "Unsupported source type unit: " + sourceTypeUnit);
```

Comment:
Style nit: Instead of throwing the exception three times, it is possible to just do it once after the outer case statement.

Reply:
I double-checked these; at least as they stand now, they are independent error cases with independent error messages.